kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt

```diff
@@ -84,11 +110,23 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
 
             aggregateEntity.setUnindexedProperty(eventsProperty, aggregateEvents)
             aggregateEntity.setUnindexedProperty(versionProperty, currentVersion + events.size)
-            dataStore.put(transaction, listOf(snapshot, aggregateEntity))
-
-            transaction.commit()
+            return try {
+                dataStore.put(transaction, listOf(snapshotEntity, aggregateEntity))
+                transaction.commit()
+                SaveEventsResponse.Success(aggregateId, currentVersion)
+            } catch (ex: ApiProxy.RequestTooLargeException) {
+                var snapshot: Snapshot? = null
+                //if a build snapshot does not exist it would not have the field data filled in.
+                if (snapshotEntity.getProperty("data") as Blob? != null) {
```

Comment:
This code looks really strange as logic to me. 

1. HttpEventStore Client -> RPC -> AppEngineEventStore -> LimitReachedException
2. HttpEventStore Client (makes snapshot of the current aggregate) -> RPC -> AppEngineEventStore  -> creates snapshot add stores new events in the new datastore

This is how I understand the flow.

Reply:
The workflow is as follows The aggregate repository tries to save an entity (e.g add the new events) 
the EventStore returns that it has reached it's limit with an SaveEventsResponse.CurrentEventsLimitReached
```Kotlin 
data class CurrentEventsLimitReached(val currentEvents: List<EventPayload>, val currentSnapshot: Snapshot? = null) : SaveEventsResponse()
```
As i see the object is build from the events in the SimpleAggregateRepository, that's why i've chosen to return the current events and a snapshot if available back to the SimpleAggregateRepository. 
The fact that the current events limit has been reached prompts the SimpleAggregateRepository to retry it's request with the same SaveOptions but including a new setting 

```Kotlin
data class CreateSnapshot(val required: Boolean = false, val snapshotData: String = "")
```
with required set to true and the newly build snapshot.

Could be optimized with the snapshotData being a byte array altogether. 

The new snapshot is being build from the old snapshot ( if available) and the current events that have reached their limit but without the new ones. This way the new events will not by digested in the new snapshot and will be present in the new entity ( or whatever persistence we have chosen)

Then sending the new snapshot with the events that caused the event limit reach and saving them in a new entity. 

This way the SimpleAggregateRepository retries the save of the events by itself and manages the snapshot creation.